sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+            if (StorageHelper.PreferencesStorage.getString(PREFERENCE_KEY_DOWNLOAD_URI) == null) {
+                mWorkflowCompleted = false;
+                mBrowserOpened = false;
+            }
+        }
+    }
+
+    @Override
+    public synchronized void onActivityResumed(Activity activity) {
+        mForegroundActivity = activity;
+        resumeUpdateWorkflow();
+    }
+
+    @Override
+    public synchronized void onActivityPaused(Activity activity) {
+        mForegroundActivity = null;
+    }
+
+    @Override
+    public synchronized void setInstanceEnabled(boolean enabled) {
+        super.setInstanceEnabled(enabled);
+        if (enabled) {
+            resumeUpdateWorkflow();
+        } else {
+
+            /* Clean all state on disabling, cancel everything. */
+            mBrowserOpened = false;
+            mWorkflowCompleted = false;
+            cancelPreviousTasks();
+            StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_UPDATE_TOKEN);
+            StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_REQUEST_ID);
+        }
+    }
+
+    /**
+     * Implements {@link #setLoginUrl(String)}.
+     */
+    private synchronized void setInstanceLoginUrl(String loginUrl) {
+        mLoginUrl = loginUrl;
+    }
+
+    /**
+     * Implements {@link #setApiUrl(String)}}.
+     */
+    private synchronized void setInstanceApiUrl(String apiUrl) {
+        mApiUrl = apiUrl;
+    }
+
+    /**
+     * Cancel everything.
+     */
+    private synchronized void cancelPreviousTasks() {
+        if (mCheckReleaseApiCall != null) {
+            mCheckReleaseApiCall.cancel();
+            mCheckReleaseApiCall = null;
+            mCheckReleaseCallId = null;
+        }
+        mUpdateDialog = null;
+        mReleaseDetails = null;
+        if (mDownloadTask != null) {
+            mDownloadTask.cancel(true);
+            mDownloadTask = null;
+        }
+        if (mProcessDownloadCompletionTask != null) {
+            mProcessDownloadCompletionTask.cancel(true);
+            mProcessDownloadCompletionTask = null;
+        }
+        long downloadId = StorageHelper.PreferencesStorage.getLong(PREFERENCE_KEY_DOWNLOAD_ID);
+        if (downloadId > 0) {
+            MobileCenterLog.debug(LOG_TAG, "Removing download and notification id=" + downloadId);
+            removeDownload(downloadId);
+        }
+        StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_DOWNLOAD_ID);
+        StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_DOWNLOAD_URI);
+    }
+
+    /**
+     * Method that triggers the update workflow or proceed to the next step.
+     */
+    private synchronized void resumeUpdateWorkflow() {
+        if (mForegroundActivity != null && !mWorkflowCompleted && isInstanceEnabled()) {
+
+            /* If we received the update token before Mobile Center was started/enabled, process it now. */
+            if (mBeforeStartUpdateToken != null) {
```

Comment:
Updates.isEnabled will throw null pointer exception is you don't initilize storage. Do you really want to do it that way? DeepLinkActivity will first tell Updates to store token in memory then the code resume the app which in the case application is killed will start the launcher that will call onStarted then onActivityResume